Sync/Database/InsertContactsQuery.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System.Collections.Generic;
+using System.Data.SqlClient;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Sync.Database
+{
+    internal class InsertContactsQuery : IQuery
+    {
+        private const string queryMergeStatement = "MERGE INTO dbo.Contacts AS target USING (VALUES ";
+        private const string queryUpdateStatment = @") AS source (Id, Name, ContactType, ContactName, Address, Email, Phone)
+ON target.Id = source.Id
+WHEN MATCHED THEN 
+    UPDATE SET 
+        Name = source.Name,
+        ContactType = source.ContactType,
+        ContactName = source.ContactName,
+        Address = source.Address,
+        Email = source.Email,
+        Phone = source.Phone
+WHEN NOT MATCHED THEN
+    INSERT (Id, Name, ContactType, ContactName, Address, Email, Phone)
+    VALUES (source.Id, source.Name, source.ContactType, source.ContactName, source.Address, source.Email, source.Phone);";
+      
+        private readonly IEnumerable<AbbreviatedContact> contacts;
+        private readonly string connectionString;
+
+        public InsertContactsQuery(IEnumerable<AbbreviatedContact> contacts, IConfiguration config) 
+        {
+            this.contacts = contacts;
+            this.connectionString = config.GetValue("DBConnectionString");
+        }
+
+        public async Task Execute()
+        {
+            var queryCommands = CreateSqlCommands();
+            using (SqlConnection conn = new SqlConnection(connectionString))
+            {
+                try
+                {
+                    conn.Open();
+                    foreach (var command in queryCommands)
+                    {
+                        command.Connection = conn;
+                        await command.ExecuteNonQueryAsync();
+                    }
+                }
+                finally
+                {
+                    conn.Close();
+                }
+            }
+        }
+
+        private List<SqlCommand> CreateSqlCommands()
+        {
+            var commands = new List<SqlCommand>();
+            var commandGroups = BatchContacts(contacts);
```

Comment:
With large data sets I ran into the upper limit of how many params could be added to a single query, so I batched them into groups of 200.